subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/NamespaceId.java

```diff
@@ -28,11 +36,29 @@ public class NamespaceId implements Serializable {
     private String namespace;
     private String name;
 
+    public static NamespaceId decode(String encoding) {
+        byte[] data = Base64.getDecoder().decode(encoding);
+        try (ByteArrayInputStream bais = new ByteArrayInputStream(data); ObjectInputStream ois = new ObjectInputStream(bais)) {
+            return Objects.requireNonNull(Cast.uncheckedCast(ois.readObject()));
+        } catch (Exception e) {
+            throw new RuntimeException("Failed decoding namespace ID");
+        }
+    }
+
     public NamespaceId(String namespace, String name) {
         this.namespace = namespace;
         this.name = name;
     }
 
+    public String encode() {
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream(); ObjectOutputStream oos = new ObjectOutputStream(baos)) {
```

Comment:
Why are we relying on Java serialization here? I was under the impression it is only usually used as a last resort. Can we use an `Encoder` and `Decoder` with the `writeString` and `readString` methods instead?

Reply:
> Hanlon's Razor is an adage that says, “Never attribute to malice that which can be adequately explained by stupidity.”